migration-engine/connectors/migration-connector/src/steps.rs

```diff
@@ -37,368 +35,266 @@ where
 }
 
-impl IndexStep for CreateIndex {
-    fn applies_to_index(&self, index_definition: &IndexDefinition) -> bool {
-        self.tpe == index_definition.tpe && self.fields == index_definition.fields
+impl Argument {
+    fn matches_ast_argument(&self, argument: &ast::Argument) -> bool {
+        self.name == argument.name.name && self.value == MigrationExpression::from_ast_expression(&argument.value)
     }
 }
 
-impl IndexStep for DeleteIndex {
-    fn applies_to_index(&self, index_definition: &IndexDefinition) -> bool {
-        self.tpe == index_definition.tpe && self.fields == index_definition.fields
+impl From<&ast::Argument> for Argument {
+    fn from(arg: &ast::Argument) -> Self {
+        Argument {
+            name: arg.name.name.clone(),
+            value: MigrationExpression::from_ast_expression(&arg.value),
+        }
     }
 }
 
-impl IndexStep for UpdateIndex {
-    fn applies_to_index(&self, index_definition: &IndexDefinition) -> bool {
-        self.tpe == index_definition.tpe && self.fields == index_definition.fields
+impl Into<ast::Argument> for &Argument {
+    fn into(self) -> ast::Argument {
+        ast::Argument {
+            name: ast::Identifier::new(&self.name),
+            value: self.value.to_ast_expression(),
+            span: ast::Span::empty(),
+        }
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-
-    #[test]
-    fn delete_index_must_apply_to_the_right_indexes() {
-        let definition = IndexDefinition {
-            fields: vec!["testColumn".into()],
-            tpe: IndexType::Unique,
-            name: None,
-        };
-        let correct_delete_index = DeleteIndex {
-            model: "ignored".into(),
-            fields: vec!["testColumn".into()],
-            tpe: IndexType::Unique,
-            name: None,
-        };
-
-        assert!(correct_delete_index.applies_to_index(&definition));
-
-        let delete_index = DeleteIndex {
-            tpe: IndexType::Normal,
-            ..correct_delete_index.clone()
-        };
-
-        // tpe does not match
-        assert!(!delete_index.applies_to_index(&definition));
-
-        let delete_index = DeleteIndex {
-            name: Some("index_on_testColumn".to_owned()),
-            ..correct_delete_index.clone()
-        };
-
-        // name does not match, but it does not matter
-        assert!(delete_index.applies_to_index(&definition));
+#[derive(Debug, Deserialize, Serialize, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct DirectiveLocator {
+    #[serde(flatten)]
+    pub location: DirectiveLocation,
+    pub r#directive: String,
+    /// The arguments of the directive are required to match directives that can be repeated,
+    /// like `@@unique` on a model. This is `None` when matching can be done without comparing
+    /// the arguments, and `Some` when a directive should be matched exactly.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub arguments: Option<Vec<Argument>>,
```

Comment:
Alternative solutions I considered to convey what directives need name-only matching and what directives should be matched exactly:

- An enum like `enum DirectiveMatcher { Regular(name), Repeated(name, arguments) }`
- Split the steps into `DeleteDirective` and `DeleteRepeatedDirective`, etc.

The `Option` solution could be confusing, but it seemed like the less noisy one.

I'm also sad that this is the only case where we have an array of objects in a migration step, but I couldn't think of an alternative, because we need the arguments inside the steps for repeated directives. Otherwise matching breaks.

Reply:
I agree that it's confusing. I'll try to come up with other names and discuss it, but otherwise your suggestion sounds good!

Reply:
I just renamed to what you suggested. I thought of `DirectiveLocator => DirectiveMatcher` too but your suggestion is nice.